File: zenscript-code-model/src/main/java/raylras/zen/model/resolve/TypeResolver.java
package raylras.zen.model.resolve;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.RuleNode;
import raylras.zen.model.CompilationUnit;
import raylras.zen.model.Compilations;
import raylras.zen.model.Visitor;
import raylras.zen.model.symbol.*;
import raylras.zen.model.type.*;
import raylras.zen.util.CSTNodes;
import raylras.zen.util.Executables;
import raylras.zen.util.Operators;
import raylras.zen.util.Symbols;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static raylras.zen.model.parser.ZenScriptParser.*;

public final class TypeResolver {

    private TypeResolver() {}

    public static Optional<Type> getType(ParseTree cst, CompilationUnit unit) {
        return Optional.ofNullable(cst)
                .map(it -> it.accept(new TypeVisitor(unit)));
    }

    private static final class TypeVisitor extends Visitor<Type> {
        final CompilationUnit unit;

        TypeVisitor(CompilationUnit unit) {
            this.unit = unit;
        }

        @Override
        public Type visitImportDeclaration(ImportDeclarationContext ctx) {
            return unit.getSymbol(ctx, ImportSymbol.class)
                    .map(Symbol::getType)
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitFunctionDeclaration(FunctionDeclarationContext ctx) {
            List<Type> paramTypes = toTypeList(ctx.formalParameterList());
            Type returnType = visit(ctx.returnType());
            if (returnType == null) {
                returnType = AnyType.INSTANCE;
            }
            return new FunctionType(returnType, paramTypes);
        }

        @Override
        public Type visitExpandFunctionDeclaration(ExpandFunctionDeclarationContext ctx) {
            List<Type> paramTypes = toTypeList(ctx.formalParameterList());
            Type returnType = visit(ctx.returnType());
            if (returnType == null) {
                returnType = AnyType.INSTANCE;
            }
            return new FunctionType(returnType, paramTypes);
        }

        @Override
        public Type visitFormalParameter(FormalParameterContext ctx) {
            if (ctx.typeLiteral() != null) {
                return visit(ctx.typeLiteral());
            }
            if (ctx.defaultValue() != null) {
                return visit(ctx.defaultValue());
            }
            FormalParameterListContext parameterList = (FormalParameterListContext) ctx.getParent();
            FunctionExprContext functionExpr = ((FunctionExprContext) parameterList.getParent());
            int argumentIndex = parameterList.formalParameter().indexOf(ctx);
            Type type = visit(functionExpr);
            if (type instanceof FunctionType functionType) {
                return functionType.parameterTypes().get(argumentIndex);
            } else if (type instanceof ClassType classType) {
                return Symbols.getAnonymousFunction(classType, unit.getEnv())
                        .map(it -> it.getParameterList().get(argumentIndex))
                        .map(ParameterSymbol::getType)
                        .orElse(AnyType.INSTANCE);
            }
            return AnyType.INSTANCE;
        }

        @Override
        public Type visitDefaultValue(DefaultValueContext ctx) {
            return visit(ctx.expression());
        }

        @Override
        public Type visitReturnType(ReturnTypeContext ctx) {
            return visit(ctx.typeLiteral());
        }

        @Override
        public Type visitClassDeclaration(ClassDeclarationContext ctx) {
            return unit.getSymbol(ctx, ClassSymbol.class)
                    .map(Symbol::getType)
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitConstructorDeclaration(ConstructorDeclarationContext ctx) {
            List<Type> paramTypes = toTypeList(ctx.formalParameterList());
            Type returnType;
            if (unit.getSymbol(ctx).orElse(null) instanceof ConstructorSymbol symbol) {
                returnType = symbol.getDeclaringClass().getType();
            } else {
                returnType = AnyType.INSTANCE;
            }
            return new FunctionType(returnType, paramTypes);
        }

        @Override
        public Type visitVariableDeclaration(VariableDeclarationContext ctx) {
            if (ctx.typeLiteral() != null) {
                return visit(ctx.typeLiteral());
            } else {
                return visit(ctx.initializer());
            }
        }

        @Override
        public Type visitInitializer(InitializerContext ctx) {
            return visit(ctx.expression());
        }

        @Override
        public Type visitOperatorFunctionDeclaration(OperatorFunctionDeclarationContext ctx) {
            List<Type> paramTypes = toTypeList(ctx.formalParameterList());
            Type returnType = visit(ctx.returnType());
            return new FunctionType(returnType, paramTypes);
        }

        @Override
        public Type visitForeachVariable(ForeachVariableContext ctx) {
            // variable -> variableList -> forEach
            ForeachStatementContext forEachStatement = (ForeachStatementContext) ctx.getParent().getParent();
            List<ForeachVariableContext> variables = forEachStatement.foreachVariableList().foreachVariable();
            Type iterableType = visit(forEachStatement.expression());
            Type result = Operators.apply(iterableType, Operator.FOR_IN, unit.getEnv()).orElse(null);
            if (result instanceof ListType listType) {
                return getListForeachVariableType(listType.elementType(), ctx, variables);
            }
            if (result instanceof MapType mapType) {
                return getMapForeachVariableType(mapType, ctx, variables);
            }
            return AnyType.INSTANCE;
        }

        @Override
        public Type visitAssignmentExpr(AssignmentExprContext ctx) {
            return visit(ctx.left);
        }

        @Override
        public Type visitThisExpr(ThisExprContext ctx) {
            return lookupLocalSymbol(ctx, "this")
                    .map(Symbol::getType)
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitMapLiteralExpr(MapLiteralExprContext ctx) {
            Optional<MapEntryContext> firstEntry = Optional.ofNullable(ctx.mapEntryList())
                    .map(entries -> entries.mapEntry(0));
            Type keyType = firstEntry.map(entry -> visit(entry.key)).orElse(AnyType.INSTANCE);
            Type valueType = firstEntry.map(entry -> visit(entry.value)).orElse(AnyType.INSTANCE);
            return new MapType(keyType, valueType);
        }

        @Override
        public Type visitIntRangeExpr(IntRangeExprContext ctx) {
            return Operators.apply(visit(ctx.from), visit(ctx.to), Operator.RANGE, unit.getEnv()).orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitSimpleNameExpr(SimpleNameExprContext ctx) {
            return lookupSymbol(ctx, ctx.getText())
                    .map(Symbol::getType)
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitBinaryExpr(BinaryExprContext ctx) {
            String literal = ctx.op.getText();
            Type leftType = visit(ctx.left);
            Type rightType = visit(ctx.right);
            return Operator.of(literal, Operator.Kind.BINARY)
                    .flatMap(op -> Operators.apply(leftType, rightType, op, unit.getEnv()))
                    .orElse(leftType);
        }

        @Override
        public Type visitParensExpr(ParensExprContext ctx) {
            return visit(ctx.expression());
        }

        @Override
        public Type visitTypeCastExpr(TypeCastExprContext ctx) {
            return visit(ctx.typeLiteral());
        }

        @Override
        public Type visitFunctionExpr(FunctionExprContext ctx) {
            if (ctx.typeLiteral() != null) {
                return visit(ctx.typeLiteral());
            } else {
                // functionExpr -> assignExpr|callExpr
                ParserRuleContext caller = ctx.getParent();
                if (caller instanceof AssignmentExprContext) {
                    Type leftType = visit(caller);
                    if (leftType != null) {
                        return leftType;
                    }
                } else if (caller.getParent() instanceof CallExprContext callExpr) {
                    ExpressionContext expression = callExpr.expression();
                    if (expression instanceof MemberAccessExprContext memberAccessExpr) {
                        List<Type> argumentTypes = new ArrayList<>();
                        List<ExpressionContext> callExpressions = callExpr.expressionList().expression();
                        int functionExprPosition = callExpressions.indexOf(ctx);
                        for (int i = 0; i < functionExprPosition; i++) {
                            Type argumentType = visit(callExpressions.get(i));
                            if (argumentType == null) {
                                argumentType = AnyType.INSTANCE;
                            }
                            argumentTypes.add(argumentType);
                        }
                        Type type = visit(memberAccessExpr.expression());
                        String name = memberAccessExpr.simpleName().getText();
                        List<Executable> functions = Symbols.getExecutableMembersByName(type, name, unit.getEnv()).toList();
                        return Executables.predictNextArgumentType(functions, argumentTypes, unit.getEnv());
                    }
                }
            }
            List<Type> paramTypes = new ArrayList<>();
            for (int i = 0; i < ctx.formalParameterList().formalParameter().size(); i++) {
                paramTypes.add(AnyType.INSTANCE);
            }
            return new FunctionType(AnyType.INSTANCE, paramTypes);
        }

        @Override
        public Type visitBracketHandlerExpr(BracketHandlerExprContext ctx) {
            // FIXME
            return AnyType.INSTANCE;
        }

        @Override
        public Type visitUnaryExpr(UnaryExprContext ctx) {
            Type type = visit(ctx.expression());
            return Operator.of(ctx.op.getText(), Operator.Kind.UNARY)
                    .flatMap(op -> Operators.apply(type, op, unit.getEnv()))
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitTernaryExpr(TernaryExprContext ctx) {
            return visit(ctx.truePart);
        }

        @Override
        public Type visitLiteralExpr(LiteralExprContext ctx) {
            return switch (CSTNodes.getTokenType(ctx.literal)) {
                case DECIMAL_LITERAL -> {
                    String literal = ctx.literal.getText();
                    char last = literal.charAt(literal.length() - 1);
                    yield switch (last) {
                        case 'l', 'L' -> LongType.INSTANCE;
                        default -> IntType.INSTANCE;
                    };
                }
                case FLOAT_LITERAL -> {
                    String literal = ctx.literal.getText();
                    char last = literal.charAt(literal.length() - 1);
                    yield switch (last) {
                        case 'f', 'F' -> FloatType.INSTANCE;
                        default -> DoubleType.INSTANCE;
                    };
                }
                case HEX_LITERAL -> IntType.INSTANCE;
                case STRING_LITERAL -> StringType.INSTANCE;
                case TRUE, FALSE -> BoolType.INSTANCE;
                case NULL -> AnyType.INSTANCE;
                default -> null;
            };
        }

        @Override
        public Type visitLogicExpr(LogicExprContext ctx) {
            return visit(ctx.left);
        }

        @Override
        public Type visitMemberAccessExpr(MemberAccessExprContext ctx) {
            Type leftType = visit(ctx.expression());
            if (!(leftType instanceof SymbolProvider provider)) {
                return AnyType.INSTANCE;
            }
            if (ctx.simpleName() == null) {
                return AnyType.INSTANCE;
            }
            String simpleName = ctx.simpleName().getText();
            for (Symbol member : provider.withExpands(unit.getEnv()).getSymbols()) {
                if (member.getName().equals(simpleName)) {
                    return member.getType();
                }
            }
            return Operators.apply(leftType, StringType.INSTANCE, Operator.MEMBER_GET, unit.getEnv()).orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitArrayLiteralExpr(ArrayLiteralExprContext ctx) {
            Type firstElementType = visit(ctx.expressionList().expression(0));
            if (firstElementType == null) {
                firstElementType = AnyType.INSTANCE;
            }
            return new ArrayType(firstElementType);
        }

        @Override
        public Type visitCallExpr(CallExprContext ctx) {
            if (ctx.expression() instanceof MemberAccessExprContext memberAccessExpr) {
                Type owner = visit(memberAccessExpr.expression());
                if (owner == null) {
                    return null;
                }
                List<Type> argumentTypes = new ArrayList<>();
                for (ExpressionContext expressionContext : ctx.expressionList().expression()) {
                    Type argumentType = visit(expressionContext);
                    if (argumentType == null) {
                        argumentType = AnyType.INSTANCE;
                    }
                    argumentTypes.add(argumentType);
                }
                List<Executable> functions = Symbols.getExecutableMembersByName(owner, memberAccessExpr.simpleName().getText(), unit.getEnv()).toList();
                Executable matchedFunction = Executables.findBestMatch(functions, argumentTypes, unit.getEnv());
                return matchedFunction == null ? null : matchedFunction.getReturnType();
            } else {
                Type leftType = visit(ctx.expression());
                if (leftType instanceof FunctionType functionType) {
                    return functionType.returnType();
                } else if (leftType instanceof ClassType) {
                    return leftType;
                } else {
                    return null;
                }
            }
        }

        @Override
        public Type visitMemberIndexExpr(MemberIndexExprContext ctx) {
            Type leftType = visit(ctx.left);
            Type rightType = visit(ctx.index);
            return Operators.apply(leftType, rightType, Operator.INDEX_GET, unit.getEnv())
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitCompareExpr(CompareExprContext ctx) {
            Type leftType = visit(ctx.left);
            Type rightType = visit(ctx.right);
            return Operator.of(ctx.op.getText(), Operator.Kind.BINARY)
                    .flatMap(op -> Operators.apply(leftType, rightType, op, unit.getEnv()))
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visitArrayType(ArrayTypeContext ctx) {
            Type elementType = visit(ctx.typeLiteral());
            return new ArrayType(elementType);
        }

        @Override
        public Type visitIntersectionType(IntersectionTypeContext ctx) {
            List<Type> types = ctx.typeLiteral().stream()
                    .map(this::visit)
                    .toList();
            return new IntersectionType(types);
        }

        @Override
        public Type visitMapType(MapTypeContext ctx) {
            Type keyType = visit(ctx.key);
            Type valueType = visit(ctx.value);
            return new MapType(keyType, valueType);
        }

        @Override
        public Type visitFunctionType(FunctionTypeContext ctx) {
            List<Type> paramTypes = toTypeList(ctx.typeLiteralList());
            Type returnType = visitReturnType(ctx.returnType());
            return new FunctionType(returnType, paramTypes);
        }

        @Override
        public Type visitListType(ListTypeContext ctx) {
            Type elementType = visit(ctx.typeLiteral());
            return new ListType(elementType);
        }

        @Override
        public Type visitPrimitiveType(PrimitiveTypeContext ctx) {
            return switch (CSTNodes.getTokenType(ctx.start)) {
                case ANY -> AnyType.INSTANCE;
                case BYTE -> ByteType.INSTANCE;
                case SHORT -> ShortType.INSTANCE;
                case INT -> IntType.INSTANCE;
                case LONG -> LongType.INSTANCE;
                case FLOAT -> FloatType.INSTANCE;
                case DOUBLE -> DoubleType.INSTANCE;
                case BOOL -> BoolType.INSTANCE;
                case VOID -> VoidType.INSTANCE;
                case STRING -> StringType.INSTANCE;
                default -> null;
            };
        }

        @Override
        public Type visitClassType(ClassTypeContext ctx) {
            return SymbolResolver.lookupClass(ctx.qualifiedName(), unit).stream()
                    .findFirst()
                    .map(Symbol::getType)
                    .orElse(AnyType.INSTANCE);
        }

        @Override
        public Type visit(ParseTree node) {
            if (node != null) {
                return node.accept(this);
            } else {
                return null;
            }
        }

        @Override
        public Type visitChildren(RuleNode node) {
            return null;
        }

        Optional<Symbol> lookupSymbol(ParseTree cst, String simpleName) {
            return lookupLocalSymbol(cst, simpleName)
                    .or(() -> lookupGlobalSymbol(simpleName));
        }

        Optional<Symbol> lookupLocalSymbol(ParseTree cst, String simpleName) {
            return Compilations.lookupScope(unit, cst)
                    .map(scope -> scope.lookupSymbol(simpleName));
        }

        Optional<Symbol> lookupGlobalSymbol(String simpleName) {
            return unit.getEnv().getGlobals()
                    .filter(symbol -> symbol.getName().equals(simpleName))
                    .findFirst();
        }

        List<Type> toTypeList(FormalParameterListContext ctx) {
            return ctx.formalParameter().stream().map(this::visit).toList();
        }

        List<Type> toTypeList(TypeLiteralListContext ctx) {
            return ctx.typeLiteral().stream().map(this::visit).toList();
        }

        Type getListForeachVariableType(Type elementType, ForeachVariableContext variable, List<ForeachVariableContext> variables) {
            int total = variables.size();
            int index = variables.indexOf(variable);
            if (total == 1) {
                return elementType;
            }
            if (total == 2) {
                return index == 0 ? IntType.INSTANCE : elementType;
            }
            return null;
        }

        Type getMapForeachVariableType(MapType mapType, ForeachVariableContext variable, List<ForeachVariableContext> variables) {
            int total = variables.size();
            int index = variables.indexOf(variable);
            if (total == 1) {
                return mapType.keyType();
            }
            if (total == 2) {
                return index == 0 ? mapType.keyType() : mapType.valueType();
            }
            return null;
        }
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/resolve/TypeResolver.kt
package raylras.zen.model.resolve

import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.RuleNode
import raylras.zen.model.CompilationUnit
import raylras.zen.model.Visitor
import raylras.zen.model.lookupScope
import raylras.zen.model.parser.ZenScriptParser.*
import raylras.zen.model.symbol.*
import raylras.zen.model.type.*

fun getType(cst: ParseTree?, unit: CompilationUnit): Type {
    return cst?.accept(TypeVisitor(unit)) ?: ErrorType
}

private class TypeVisitor(val unit: CompilationUnit) : Visitor<Type>() {
    override fun visitImportDeclaration(ctx: ImportDeclarationContext): Type {
        return unit.symbolMap[ctx]?.type ?: AnyType
    }

    override fun visitFunctionDeclaration(ctx: FunctionDeclarationContext): Type {
        val paramTypes: List<Type> = mapToType(ctx.formalParameter())
        val returnType: Type = visit(ctx.returnType())
        return FunctionType(returnType, paramTypes)
    }

    override fun visitExpandFunctionDeclaration(ctx: ExpandFunctionDeclarationContext): Type {
        val paramTypes: List<Type> = mapToType(ctx.formalParameter())
        val returnType: Type = visit(ctx.returnType())
        return FunctionType(returnType, paramTypes)
    }

    override fun visitFormalParameter(ctx: FormalParameterContext): Type {
        when {
            ctx.typeLiteral() != null -> {
                return visit(ctx.typeLiteral())
            }

            ctx.defaultValue() != null -> {
                return visit(ctx.defaultValue())
            }

            else -> {
                val functionExpr = ctx.getParent() as? FunctionExprContext ?: return AnyType
                val index = functionExpr.formalParameter().indexOf(ctx)
                return when (val type = visit(functionExpr)) {
                    is FunctionType -> {
                        type.parameterTypes[index]
                    }

                    is ClassType -> {
                        type.firstAnonymousFunctionOrNull()
                            ?.parameters?.get(index)?.type
                            ?: AnyType
                    }

                    else -> AnyType
                }
            }
        }
    }

    override fun visitDefaultValue(ctx: DefaultValueContext): Type {
        return visit(ctx.expression())
    }

    override fun visitReturnType(ctx: ReturnTypeContext): Type {
        return visit(ctx.typeLiteral())
    }

    override fun visitClassDeclaration(ctx: ClassDeclarationContext): Type {
        return unit.symbolMap[ctx]?.type ?: AnyType
    }

    override fun visitConstructorDeclaration(ctx: ConstructorDeclarationContext): Type {
        val paramTypes: List<Type> = mapToType(ctx.formalParameter())
        val returnType: Type = when (val symbol = unit.symbolMap[ctx]) {
            is ConstructorSymbol -> symbol.declaringClass.type
            else -> AnyType
        }
        return FunctionType(returnType, paramTypes)
    }

    override fun visitVariableDeclaration(ctx: VariableDeclarationContext): Type {
        return when {
            ctx.typeLiteral() != null -> {
                visit(ctx.typeLiteral())
            }

            else -> {
                visit(ctx.initializer)
            }
        }
    }

    override fun visitOperatorFunctionDeclaration(ctx: OperatorFunctionDeclarationContext): Type {
        val paramTypes: List<Type> = mapToType(ctx.formalParameter())
        val returnType = visit(ctx.returnType())
        return FunctionType(returnType, paramTypes)
    }

    override fun visitForeachVariable(ctx: ForeachVariableContext): Type {
        val statement = ctx.getParent() as? ForeachStatementContext ?: return ErrorType
        val variables = statement.foreachVariable()
        val expr = visit(statement.expression())
        return when (val type = expr.applyUnaryOperator(Operator.FOR_IN, unit.env)) {
            is ListType -> {
                when (variables.size) {
                    // for v in expr
                    1 -> type.elementType

                    // for i, v in expr
                    2 -> {
                        when (variables.indexOf(ctx)) {
                            // case i
                            0 -> IntType

                            // case v
                            1 -> type.elementType

                            else -> ErrorType
                        }
                    }

                    else -> ErrorType
                }
            }

            is MapType -> {
                when (variables.size) {
                    // for v in expr
                    1 -> type.valueType

                    // for k, v in expr
                    2 -> {
                        when (variables.indexOf(ctx)) {
                            // case k
                            0 -> type.keyType

                            // case v
                            1 -> type.valueType

                            else -> ErrorType
                        }
                    }

                    else -> ErrorType
                }
            }

            else -> AnyType
        }
    }

    override fun visitAssignmentExpr(ctx: AssignmentExprContext): Type {
        return visit(ctx.left)
    }

    override fun visitThisExpr(ctx: ThisExprContext): Type {
        return lookupLocalSymbol(ctx, "this")
            ?.type
            ?: AnyType
    }

    override fun visitMapLiteralExpr(ctx: MapLiteralExprContext): Type {
        val firstEntry = ctx.mapEntryList()?.mapEntry()?.firstOrNull()
        val keyType = firstEntry?.let { visit(it.key) } ?: AnyType
        val valueType = firstEntry?.let { visit(it.value) } ?: AnyType
        return MapType(keyType, valueType)
    }

    override fun visitIntRangeExpr(ctx: IntRangeExprContext): Type {
        val from = visit(ctx.from)
        val to = visit(ctx.to)
        return from.applyBinaryOperator(Operator.RANGE, to, unit.env) ?: AnyType
    }

    override fun visitSimpleNameExpr(ctx: SimpleNameExprContext): Type {
        return lookupSymbol(ctx, ctx.getText())?.type ?: AnyType
    }

    override fun visitBinaryExpr(ctx: BinaryExprContext): Type {
        val leftType = visit(ctx.left)
        val rightType = visit(ctx.right)
        val op = Operator.of(ctx.op.text, Operator.Kind.BINARY)
        return leftType.applyBinaryOperator(op, rightType, unit.env) ?: leftType
    }

    override fun visitParensExpr(ctx: ParensExprContext): Type {
        return visit(ctx.expression())
    }

    override fun visitTypeCastExpr(ctx: TypeCastExprContext): Type {
        return visit(ctx.typeLiteral())
    }

    override fun visitFunctionExpr(ctx: FunctionExprContext): Type {
        return when {
            ctx.typeLiteral() != null -> {
                visit(ctx.typeLiteral())
            }

            else -> {
                // FIXME
                AnyType
            }
        }
    }

    override fun visitBracketHandlerExpr(ctx: BracketHandlerExprContext): Type {
        // FIXME
        return AnyType
    }

    override fun visitUnaryExpr(ctx: UnaryExprContext): Type {
        val type = visit(ctx.expression())
        val op = Operator.of(ctx.op.text, Operator.Kind.UNARY)
        return type.applyUnaryOperator(op, unit.env) ?: AnyType
    }

    override fun visitTernaryExpr(ctx: TernaryExprContext): Type {
        return visit(ctx.truePart)
    }

    override fun visitLiteralExpr(ctx: LiteralExprContext): Type {
        return when (ctx.literal?.type) {
            DECIMAL_LITERAL -> {
                when (ctx.literal.text.last()) {
                    'l', 'L' -> LongType
                    else -> IntType
                }
            }

            FLOAT_LITERAL -> {
                when (ctx.literal.text.last()) {
                    'f', 'F' -> FloatType
                    else -> DoubleType
                }
            }

            HEX_LITERAL -> IntType
            STRING_LITERAL -> StringType
            TRUE, FALSE -> BoolType
            NULL -> AnyType
            else -> AnyType
        }
    }

    override fun visitLogicExpr(ctx: LogicExprContext): Type {
        return visit(ctx.left)
    }

    override fun visitMemberAccessExpr(ctx: MemberAccessExprContext): Type {
        val leftType = visit(ctx.expression())
        when {
            leftType !is SymbolProvider -> {
                return AnyType
            }

            ctx.simpleName() == null -> {
                return AnyType
            }

            else -> {
                leftType.getSymbols(unit.env)
                    .firstOrNull { it.simpleName == ctx.simpleName().text }
                    ?.let { return it.type }
                return leftType.applyBinaryOperator(Operator.MEMBER_GET, StringType, unit.env) ?: AnyType
            }
        }
    }

    override fun visitArrayLiteralExpr(ctx: ArrayLiteralExprContext): Type {
        val firstElementType: Type = visit(ctx.expressionList().expression().firstOrNull())
        return ArrayType(firstElementType)
    }

    override fun visitCallExpr(ctx: CallExprContext): Type {
        // FIXME
        return when (val caller = visit(ctx.expression())) {
            is Executable -> {
                AnyType
            }

            else -> {
                AnyType
            }
        }
    }

    override fun visitMemberIndexExpr(ctx: MemberIndexExprContext): Type {
        val leftType = visit(ctx.left)
        val rightType = visit(ctx.index)
        return leftType.applyBinaryOperator(Operator.INDEX_GET, rightType, unit.env) ?: AnyType
    }

    override fun visitCompareExpr(ctx: CompareExprContext): Type {
        val leftType = visit(ctx.left)
        val rightType = visit(ctx.right)
        val op = Operator.of(ctx.op.text, Operator.Kind.BINARY)
        return leftType.applyBinaryOperator(op, rightType, unit.env) ?: AnyType
    }

    override fun visitArrayType(ctx: ArrayTypeContext): Type {
        val elementType = visit(ctx.typeLiteral())
        return ArrayType(elementType)
    }

    override fun visitIntersectionType(ctx: IntersectionTypeContext): Type {
        val types = ctx.typeLiteral().map { visit(it) }
        return IntersectionType(types)
    }

    override fun visitMapType(ctx: MapTypeContext): Type {
        val keyType = visit(ctx.key)
        val valueType = visit(ctx.value)
        return MapType(keyType, valueType)
    }

    override fun visitFunctionType(ctx: FunctionTypeContext): Type {
        val paramTypes: List<Type> = mapToType(ctx.typeLiteral())
        val returnType = visitReturnType(ctx.returnType())
        return FunctionType(returnType, paramTypes)
    }

    override fun visitListType(ctx: ListTypeContext): Type {
        val elementType = visit(ctx.typeLiteral())
        return ListType(elementType)
    }

    override fun visitPrimitiveType(ctx: PrimitiveTypeContext): Type {
        return when (ctx.start?.type) {
            ANY -> AnyType
            BYTE -> ByteType
            SHORT -> ShortType
            INT -> IntType
            LONG -> LongType
            FLOAT -> FloatType
            DOUBLE -> DoubleType
            BOOL -> BoolType
            VOID -> VoidType
            STRING -> StringType
            else -> ErrorType
        }
    }

    override fun visitClassType(ctx: ClassTypeContext): Type {
        return lookupClass(ctx.qualifiedName(), unit)
            .firstOrNull()
            ?.type
            ?: AnyType
    }

    override fun visit(node: ParseTree?): Type {
        return node?.accept(this) ?: ErrorType
    }

    override fun visitChildren(node: RuleNode): Type? {
        return null
    }

    private fun lookupSymbol(cst: ParseTree?, simpleName: String): Symbol? {
        return lookupLocalSymbol(cst, simpleName) ?: lookupGlobalSymbol(simpleName)
    }

    private fun lookupLocalSymbol(cst: ParseTree?, simpleName: String): Symbol? {
        return lookupScope(cst, unit)?.firstOrNull { it.simpleName == simpleName }
    }

    private fun lookupGlobalSymbol(simpleName: String?): Symbol? {
        return unit.env.globals.firstOrNull { it.simpleName == simpleName }
    }

    private inline fun <reified T> mapToType(ctxList: List<ParseTree>): List<T> {
        return ctxList.map { visit(it) as T }
    }
}
